python/cudf/cudf/core/udf/strings_lowering.py

```diff
@@ -136,45 +142,100 @@ def cast_string_literal_to_string_view(context, builder, fromty, toty, val):
+        default_manager[managed_udf_string].get_value_type()
+    )
+    udf_str_ptr = builder.gep(
+        managed_ptr, [ir.IntType(32)(0), ir.IntType(32)(1)]
+    )
+
     builder.store(val, sv_ptr)
     _ = context.compile_internal(
         builder,
         call_create_udf_string_from_string_view,
         nb_signature(types.void, _STR_VIEW_PTR, types.CPointer(udf_string)),
         (sv_ptr, udf_str_ptr),
     )
-    result = cgutils.create_struct_proxy(udf_string)(
-        context, builder, value=builder.load(udf_str_ptr)
-    )
-
-    return result._getvalue()
+    return _finalize_new_managed_udf_string(context, builder, managed_ptr)
 
 
-@cuda_lowering_registry.lower_cast(udf_string, string_view)
-def cast_udf_string_to_string_view(context, builder, fromty, toty, val):
-    udf_str_ptr = builder.alloca(default_manager[fromty].get_value_type())
+@cuda_lowering_registry.lower_cast(managed_udf_string, string_view)
+def cast_managed_udf_string_to_string_view(
+    context, builder, fromty, toty, val
+):
     sv_ptr = builder.alloca(default_manager[toty].get_value_type())
-    builder.store(val, udf_str_ptr)
+
+    managed_ptr = builder.alloca(default_manager[fromty].get_value_type())
+    builder.store(val, managed_ptr)
+
+    udf_str_ptr = builder.gep(
+        managed_ptr, [ir.IntType(32)(0), ir.IntType(32)(1)]
+    )
 
     context.compile_internal(
         builder,
         call_create_string_view_from_udf_string,
         nb_signature(types.void, _UDF_STRING_PTR, _STR_VIEW_PTR),
         (udf_str_ptr, sv_ptr),
     )
-
     result = cgutils.create_struct_proxy(string_view)(
         context, builder, value=builder.load(sv_ptr)
     )
-
     return result._getvalue()
 
 
-# utilities
+# Utilities
+_new_meminfo_from_udf_str = cuda.declare_device(
+    "meminfo_from_new_udf_str", types.voidptr(_UDF_STRING_PTR)
+)
+
+
+def new_meminfo_from_udf_str(udf_str):
+    return _new_meminfo_from_udf_str(udf_str)
+
+
+def _finalize_new_managed_udf_string(context, builder, managed_ptr):
+    """
+    Allocate a udf_string and a NRT_MemInfo as part of one struct
+    and initialize the NRT_MemInfo with a refct=1.
+    """
+
+    # {i8*, i32, i32}*
+    udf_str_ptr = builder.gep(
+        managed_ptr, [ir.IntType(32)(0), ir.IntType(32)(1)]
+    )
+
+    # Call the shim function which initializes an NRT_MemInfo object around the
+    # udf_string pointer. The resulting pointer points to a heap allocation. A
+    # copy of the udf_string is made, although its underlying data isn't copied
+    # See shim.cu for details.
+    mi = context.compile_internal(
+        builder,
+        new_meminfo_from_udf_str,
+        types.voidptr(_UDF_STRING_PTR),
+        (udf_str_ptr,),
+    )
+
+    managed = cgutils.create_struct_proxy(managed_udf_string)(
+        context,
+        builder,
+        value=builder.load(
+            managed_ptr
+        ),  # {i8*, {i8*, i32, i32}}* -> {i8*, {i8*, i32, i32}}
+    )
+    # i8* = i8*
+    managed.meminfo = mi
```

Comment:
I'm struggling to see why we do this - doesn't this end up returning a struct that contains the same data as `mi` held anyway?